controllers/pod_node_module_reconciler.go

```diff
@@ -70,8 +72,28 @@ func (pnmr *PodNodeModuleReconciler) Reconcile(ctx context.Context, req ctrl.Req
 	if !podutils.IsPodReady(&pod) || !pod.DeletionTimestamp.IsZero() {
 		logger.Info("Unlabeling node")
 
-		if err := pnmr.deleteLabel(ctx, nodeName, labelName); err != nil {
-			return ctrl.Result{}, fmt.Errorf("could not unlabel node %s: %v", nodeName, err)
+		// Make sure we don't already have a new running pod before unlabeling the node
+		labelSelector := client.MatchingLabels{constants.ModuleNameLabel: moduleName}
+		fieldSelector := client.MatchingFields{"spec.nodeName": nodeName}
+		var modulePodsList v1.PodList
+		err := pnmr.client.List(ctx, &modulePodsList, labelSelector, fieldSelector)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to get list of all pods for module %s on node %s: %v", moduleName, nodeName, err)
+		}
+		var foundRunningPod bool
+		for _, p := range modulePodsList.Items {
+			if podutils.IsPodReady(&p) && p.DeletionTimestamp.IsZero() {
+				foundRunningPod = true
+				break
+			}
+		}
+		if !foundRunningPod {
+			if err := pnmr.deleteLabel(ctx, nodeName, labelName); err != nil {
+				return ctrl.Result{}, fmt.Errorf("could not unlabel node %s with label %q: %v", nodeName, labelName, err)
+			}
+			if err := pnmr.deleteLabel(ctx, nodeName, deprecatedLabelName); err != nil {
+				return ctrl.Result{}, fmt.Errorf("could not unlabel node %s with label %q: %v", nodeName, deprecatedLabelName, err)
+			}
```

Comment:
This applies two patches in a row. Could we apply only one with both labels?

Reply:
Good point. Will try to merge them into one patch.

Reply:
Fixed.